### DISCUSSION TITLE
zh/Q-and-A/answer/question5

### DISCUSSION THREAD

**Discussion by giscus** [vuepress-theme-hope/giscus-discussions#918] [Announcements]
# zh/Q-and-A/answer/question5

程序正确，洛谷提交错误 问题描述： 使用函数递归，程序正确，但在洛谷上提交错误，源代码（已加入必要说明）

https://cygnomatic-controller.github.io/zh/Q-and-A/answer/question5.html

**Comment by 477996850** [comment]
end() 函数的作用应该是返回一个数字是否仅由一个 2 组成。然而在循环条件里，当 right > 9 的时候，它会不断除以 10 直到 right 变成一位数。这可能导致原始的数值信息丢失，特别是如果输入的 right 非常大且只有一位包含 2。这个函数可能不会准确地判断单个的 2。

**Comment by da-zhexue** [comment]
end函数实际上起的作用是第一个数字是否是2，而judge函数在判断完end函数判断完第一个数后将最后一个数删除，所以如果一个数第一位是2，那么就会计入这个数有n个2（n表示这个数的位数）无论这个数后面是什么。但是主函数中漏计入一个left，使样例恰好过了。改的话，把end函数中改为right%10==2时，返回1即可

**Reply by da-zhexue** [reply]
还有主函数里 while (right > left)加个等号
